pkg/agent/proxy/proxier.go

```diff
@@ -355,40 +431,75 @@ func (p *proxier) installServices() {
 		if needUpdateService {
 			// Delete previous flow.
 			if needRemoval {
+				// If previous Service should be removed, remove ClusterIP flows of previous Service.
 				if err := p.ofClient.UninstallServiceFlows(pSvcInfo.ClusterIP(), uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
 					klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
 					continue
 				}
+
+				if p.proxyFullEnabled {
+					// If previous Service which has NodePort should be removed, remove NodePort flows and configurations of previous Service.
+					if pSvcInfo.NodePort() > 0 {
+						if err := p.uninstallNodePortService(uint16(pSvcInfo.NodePort()), pSvcInfo.OFProtocol); err != nil {
+							klog.Errorf("Failed to remove flows and configurations of Service %v: %v", svcPortName, err)
+							continue
+						}
+					}
+				}
 			}
+
+			// Install ClusterIP flows of current Service.
 			if err := p.ofClient.InstallServiceFlows(groupID, svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(svcInfo.StickyMaxAgeSeconds())); err != nil {
 				klog.Errorf("Error when installing Service flows: %v", err)
 				continue
 			}
-			// Install OpenFlow entries for the ingress IPs of LoadBalancer Service.
-			// The LoadBalancer Service should be accessible from Pod, Node and
-			// external host.
-			var toDelete, toAdd []string
-			if needRemoval {
-				toDelete = pSvcInfo.LoadBalancerIPStrings()
-				toAdd = svcInfo.LoadBalancerIPStrings()
-			} else {
-				toDelete = deletedLoadBalancerIPs
-				toAdd = addedLoadBalancerIPs
-			}
-			for _, ingress := range toDelete {
-				if ingress != "" {
-					// It is safe to access pSvcInfo here. If this is a new Service,
-					// then toDelete will be an empty slice.
-					if err := p.uninstallLoadBalancerServiceFlows(net.ParseIP(ingress), uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
-						klog.Errorf("Error when removing LoadBalancer Service flows: %v", err)
+
+			if p.proxyFullEnabled {
+				// Install ClusterIP route on k8s node so that ClusterIP can be accessed on k8s node. When thWhen the
+				// Service is removed, the route does not need to be removed, because the route target IP block is calculated
+				// every time a new ClusterIP is created. When there are enough ClusterIPs, the Service CIDR can be finally
```

Comment:
Maybe this explanation can be better:

```
// Install ClusterIP route on Node so that ClusterIP can be accessed on Node. Every time a new ClusterIP
// is created, the routing target IP block will be recalculated for expansion to be able to route the new
// created ClusterIP. Deleting a ClusterIP will not shrink the target routing IP block. The Service CIDR
// can be finally calculated after creating enough ClusterIPs.
```